src/lib.rs

```diff
@@ -1410,6 +1544,348 @@ where
+    ///
+    pub fn request<F, RE>(self, http_client: F) -> Result<TR, RequestTokenError<RE, TE>>
+    where
+        F: Fn(HttpRequest) -> Result<HttpResponse, RE>,
+        RE: Error + 'static,
+    {
+        let details = self.device_authorization_details.ok_or_else(|| {
+            RequestTokenError::Other("no device authorization details provided".to_string())
+        })?;
+
+        let mut elapsed = Duration::new(0, 0);
+        let mut interval = details.interval().clone();
+
+        loop {
+            if elapsed > details.expires_in() {
+                return Err(RequestTokenError::Other("Device code expired".to_string()));
+            }
+
+            // Prepare the request.
+            match http_client(self.prepare_request()?)
+                .map_err(|_| DeviceCodeAction::DoubleIntervalThenRetry)
+                .and_then(device_token_action)
+            {
+                Ok(http_response) => break token_response(http_response),
+                Err(DeviceCodeAction::Retry) => {
+                    // Wait for the current interval.
+                }
+                Err(DeviceCodeAction::IncreaseIntervalThenRetry) => {
+                    // Increase the interval, then wait that long.
+                    interval = interval + Duration::from_secs(5);
+                }
+                Err(DeviceCodeAction::DoubleIntervalThenRetry) => {
+                    // Double the interval, then wait that long.
+                    interval = interval.checked_mul(2).ok_or(RequestTokenError::Other(
+                        "Failed to increase interval".to_string(),
+                    ))?;
+                }
+                Err(DeviceCodeAction::NoFurtherRequests(req)) => break token_response(req),
+            };
+
+            // Actually sleep here.
+            std::thread::sleep(interval);
+            elapsed = elapsed + interval;
+        }
+    }
+
+    ///
+    /// Asynchronously sends the request to the authorization server and awaits a response.
+    ///
+    pub async fn request_async<C, F, RE>(
+        self,
+        http_client: C,
+    ) -> Result<TR, RequestTokenError<RE, TE>>
+    where
+        C: Fn(HttpRequest) -> F,
+        F: Future<Output = Result<HttpResponse, RE>>,
+        RE: Error + 'static,
+    {
+        let details = self.device_authorization_details.ok_or_else(|| {
+            RequestTokenError::Other("no device authorization details provided".to_string())
+        })?;
+
+        let mut elapsed = Duration::new(0, 0);
+        let mut interval = details.interval().clone();
+
+        loop {
+            if elapsed > details.expires_in() {
+                return Err(RequestTokenError::Other("Device code expired".to_string()));
+            }
+
+            // Prepare the request.
+            match http_client(self.prepare_request()?)
+                .await
+                .map_err(|_| DeviceCodeAction::DoubleIntervalThenRetry)
```

Comment:
Will have a play and see what I can come up with.

Reply:
@ramosbugs: If you could have a quick look at this comment, it'd be great :smile:

So I'm trying to implement this to stop the deduplication. Here's what I currently have:

```rust
    fn process_response<RE>(
        self,
        res: Result<HttpResponse, RE>,
        current_interval: Duration,
    ) -> DeviceAccessTokenPollResult<TR, RE, TE, TT>
    where
        RE: Error + 'static,
    {
        let http_response = match res {
            Ok(inner) => inner,
            Err(_) => {
                // Try and double the current interval. If that fails, just use the current one.
                let new_interval = current_interval.checked_mul(2).unwrap_or(current_interval);
                return DeviceAccessTokenPollResult::ContinueWithNewPollInterval(new_interval);
            }
        };

        // Explicitly process the response with a DeviceCodeErrorResponse
        let res = endpoint_response::<RE, DeviceCodeErrorResponse, TR>(http_response);
        match res {
            // On a ServerResponse error, the error needs inspecting as a DeviceCodeErrorResponse
            // to work out whether a retry needs to happen.
            Err(RequestTokenError::ServerResponse(dcer)) => {
                match dcer.error() {
                    // On AuthorizationPending, a retry needs to happen with the same poll interval.
                    DeviceCodeErrorResponseType::AuthorizationPending => {
                        DeviceAccessTokenPollResult::ContinueWithNewPollInterval(current_interval)
                    }
                    // On SlowDown, a retry needs to happen with a larger poll interval.
                    DeviceCodeErrorResponseType::SlowDown => {
                        DeviceAccessTokenPollResult::ContinueWithNewPollInterval(
                            current_interval + Duration::from_secs(5),
                        )
                    }

                    // On any other error, just return the error.
                    other => DeviceAccessTokenPollResult::Done(DeviceAccessResult::new(Err(
                        RequestTokenError::ServerResponse(dcer),
                    ))),
                }
            }

            // On any other success or failure, return the failure.
            res => DeviceAccessTokenPollResult::Done(DeviceAccessResult::new(res))),
        }
    }
```
The errors here are:
```rust
error[E0308]: mismatched types
    --> src/lib.rs:1934:59
     |
1702 | impl<'a, TE, TR, TT, EF> DeviceAccessTokenRequest<'a, TE, TR, TT, EF>
     |          -- this type parameter
...
1934 |                         RequestTokenError::ServerResponse(dcer),
     |                                                           ^^^^ expected type parameter `TE`, found struct `StandardErrorResponse`
     |
     = note: expected type parameter `TE`
                        found struct `StandardErrorResponse<devicecode::DeviceCodeErrorResponseType>`

error[E0308]: mismatched types
    --> src/lib.rs:1940:78
     |
1702 | impl<'a, TE, TR, TT, EF> DeviceAccessTokenRequest<'a, TE, TR, TT, EF>
     |          -- this type parameter
...
1940 |             res => DeviceAccessTokenPollResult::Done(DeviceAccessResult::new(res)),
     |                                                                              ^^^ expected type parameter `TE`, found struct `StandardErrorResponse`
     |
     = note: expected enum `std::result::Result<_, RequestTokenError<_, TE>>`
                found enum `std::result::Result<_, RequestTokenError<_, StandardErrorResponse<devicecode::DeviceCodeErrorResponseType>>>`
```
This is because the call to `endpoint_response` is strongly typed, because we need to get access to the error type. But because it's strongly typed, just returning it to the user won't work, because the Result is bounded on TE (which is just `TE: ErrorResponse + 'static,`, but could be any concrete type that implements these types.

So presumably I want a way of converting from `DeviceCodeErrorResponse` to `TE` - but the only generic way I can think of doing that is involving `serde_json` to Serialize the error, then Deserialize to TE, which seems nasty.

Am I just missing something obvious here?

Reply:
I think this means we should change the return type to `DeviceAccessTokenPollResult<TR, RE, DeviceCodeErrorResponse, TT>`. the other response handlers don't care about specific error response codes, so they can be fully generic. however, in this case since the API handler needs to consume some of the error codes, I guess we have to be more specific about the error type.

an alternative is similar to what you suggested: deserialize the `http_response` twice: once with `DeviceCodeErrorResponse` as the error type so that we can take the appropriate actions within this function, and then again using the generic error type to deserialize the response before returning it. I don't think it's worth that overhead or complexity, though.

a third option, which is kind of appealing, is to make `DeviceCodeErrorResponseType` generic in its type for the `Basic` variant. then `DeviceCodeErrorResponse` can be defined as:
```rust
pub type DeviceCodeErrorResponse<BE: Debug + DeserializeOwned + Serialize> = StandardErrorResponse<DeviceCodeErrorResponseType<BE>>;
```

the advantage to this approach is that it still supports extensibility for other error codes that servers might want to return that are unrelated to the device code flow.

I'm happy with either the first or third approach, so I'll leave it up to you.

Reply:
I went with the first option as I could get that working.